quinn-udp/src/windows.rs

```diff
@@ -154,4 +369,98 @@ impl UdpSocketState {
     }
 }
 
+fn set_socket_option(
+    socket: &impl AsRawSocket,
+    level: i32,
+    name: i32,
+    value: u32,
+) -> Result<(), io::Error> {
+    let rc = unsafe {
+        WinSock::setsockopt(
+            socket.as_raw_socket() as usize,
+            level,
+            name,
+            &value as *const _ as _,
+            mem::size_of_val(&value) as _,
+        )
+    };
+
+    match rc == 0 {
+        true => Ok(()),
+        false => Err(io::Error::last_os_error()),
+    }
+}
+
 pub(crate) const BATCH_SIZE: usize = 1;
+// Enough to store max(IP_PKTINFO + IP_ECN, IPV6_PKTINFO + IPV6_ECN) + UDP_SEND_MSG_SIZE bytes (header + data) and some extra margin
+const CMSG_LEN: usize = 128;
+const OPTION_ON: u32 = 1;
+
+// FIXME this could use [`std::sync::OnceLock`] once the MSRV is bumped to 1.70 and upper
+static WSARECVMSG_PTR: Lazy<WinSock::LPFN_WSARECVMSG> = Lazy::new(|| {
+    let s = unsafe { WinSock::socket(WinSock::AF_INET as _, WinSock::SOCK_DGRAM as _, 0) };
+    if s == WinSock::INVALID_SOCKET {
+        tracing::debug!(
+            "ignoring WSARecvMsg function pointer due to socket creation error: {}",
+            io::Error::last_os_error()
+        );
+        return None;
+    }
+
+    // Detect if OS expose WSARecvMsg API based on
+    // https://github.com/Azure/mio-uds-windows/blob/a3c97df82018086add96d8821edb4aa85ec1b42b/src/stdnet/ext.rs#L601
+    let guid = WinSock::WSAID_WSARECVMSG;
+    let mut wsa_recvmsg_ptr = None;
+    let mut len = 0;
+
+    // Safety: Option handles the NULL pointer with a None value
+    let rc = unsafe {
+        WinSock::WSAIoctl(
+            s as _,
+            WinSock::SIO_GET_EXTENSION_FUNCTION_POINTER,
+            &guid as *const _ as *const _,
+            mem::size_of_val(&guid) as u32,
+            &mut wsa_recvmsg_ptr as *mut _ as *mut _,
+            mem::size_of_val(&wsa_recvmsg_ptr) as u32,
+            &mut len,
+            ptr::null_mut(),
+            None,
+        )
+    };
+
+    if rc == -1 {
+        tracing::debug!(
+            "ignoring WSARecvMsg function pointer due to ioctl error: {}",
+            io::Error::last_os_error()
+        );
+    } else if len as usize != mem::size_of::<WinSock::LPFN_WSARECVMSG>() {
+        tracing::debug!("ignoring WSARecvMsg function pointer due to pointer size mismatch");
+        wsa_recvmsg_ptr = None;
+    }
+
+    unsafe {
+        WinSock::closesocket(s);
+    }
+
+    wsa_recvmsg_ptr
+});
+
+static MAX_GSO_SEGMENTS: Lazy<usize> = Lazy::new(|| {
+    let socket = match std::net::UdpSocket::bind("[::]:0")
+        .or_else(|_| std::net::UdpSocket::bind("127.0.0.1:0"))
```

Comment:
Should we use IPv4 unconditionally here? I guess maybe IPv4 could also be disabled, so we'd want to try both for robustness...

Reply:
We have same code for [linux gso detection](https://github.com/quinn-rs/quinn/blob/55234e178fdca81cd51a5bfb520cb912de14f72e/quinn-udp/src/unix.rs#L758) i think it is more robust to keep it that way.